ArduPlane/stall_detection.cpp

```diff
@@ -0,0 +1,179 @@
+ */
+
+/*
+  Do we think we are stalled?
+  Probabilistic method where a bool is low-passed and considered a probability.
+*/
+void Plane::stall_detection_update(void)
+{
+    stall_state.raw_algorithm_output = false;
+
+    if (!arming.is_armed() ||
+        auto_state.last_flying_ms == 0 ||
+        !is_flying() ||
+        crash_state.is_crashed)
+    {
+        stall_state.stall_clear();
+        stall_state.count = 0;
+        stall_detection_log();
+        return;
+    }
+
+    if (stall_state.inhibit_ms != 0) {
+        // if someone told us to inihibt, freeze all stall processing for a brief time.
+        // This is useful when changing waypoints where we will likely perform a
+        // hard turn and roll/pitch error spikes for a moment
+        if (AP_HAL::millis() - stall_state.inhibit_ms > 2000) {
+            stall_state.inhibit_ms = 0;
+        }
+        stall_detection_log();
+        return;
+    }
+
+    const bool is_stalled_initial = stall_state.is_stalled();
+
+    stall_state.raw_algorithm_output = stall_detection_algorithm(true);
+
+    // LPF confidence
+    stall_state.confidence = ((float)stall_state.raw_algorithm_output * stall_state.coef) + (stall_state.confidence * (1.0f - stall_state.coef));
+
+    if ((is_stalled_initial != stall_state.is_stalled()) && stall_state.is_stalled()) {
+        // we just stalled
+
+        gcs().send_text_rate_limited(MAV_SEVERITY_WARNING, 1000, stall_state.detect_notify_gcs_last_ms, "STALL DETECTED!");
+
+        stall_state.stall_start();
+        // TODO: trigger stall recovery
+    }
+    stall_detection_log();
+}
+
+void Plane::stall_detection_log()
+{
+    // log to AP_Logger
+    AP::logger().Write(
+       "STAL",
+       "TimeUS,Actv,Raw,Any,Conf,Inh",
+       "s-----",
+       "F-----",
+       "QBBIfI",
+       AP_HAL::micros64(),
+       (uint8_t)stall_state.is_stalled(),
+       (uint8_t)stall_state.raw_algorithm_output,
+       (uint32_t)stall_state.algorithm_output_if_everything_enabled,
+       (double)stall_state.confidence,
+       (uint32_t)(stall_state.inhibit_ms != 0) ? AP_HAL::millis() - stall_state.inhibit_ms : 0);
+}
+
+// return true if we think we're stalling
+bool Plane::stall_detection_algorithm(bool allow_changing_state)
+{
+    //
+    // Pre-process nav roll to apply rate limit. Needs to be done before potential early returns.
+    //
+
+    // Time since last update.
+    const float deltaT = 0.001f * (AP_HAL::millis() - stall_state.last_update_ms);
+
+    // Directly take nav_roll_cd, used if time since last update is too long indicating a reset.
+    float limited_nav_roll = 0.01f*nav_roll_cd;
+
+    if (deltaT < 0.5) {
+        // Difference between current nav roll and previous limited value.
+        float delta_angle      = 0.01f*nav_roll_cd - stall_state.last_limited_nav_roll;
+
+        // New limited nav roll
+        limited_nav_roll = stall_state.last_limited_nav_roll + constrain_float(delta_angle, -50.0f*deltaT, 50.0f*deltaT);
```

Comment:
@samuelctabor wrote that in, can't remember the reason now. Hope he knows!

Reply:
Yes it's the roll rate limit. The roll controller RMAX parameter should be used here. It's important that this be set accurately to avoid false positives.

It would be better if this was all done inside the roll controller, but may not be worth it depending how quickly plane moves to AC_PID.